homeassistant/components/update/__init__.py

```diff
@@ -0,0 +1,360 @@
+"""Component to allow for providing device or service updates."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+from datetime import timedelta
+import logging
+from typing import Any, Final, final
+
+import voluptuous as vol
+
+from homeassistant.backports.enum import StrEnum
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import STATE_OFF, STATE_ON
+from homeassistant.core import HomeAssistant, ServiceCall
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.config_validation import (
+    PLATFORM_SCHEMA,
+    PLATFORM_SCHEMA_BASE,
+)
+from homeassistant.helpers.entity import EntityCategory, EntityDescription
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.restore_state import RestoreEntity
+from homeassistant.helpers.typing import ConfigType
+
+from .const import (
+    ATTR_BACKUP,
+    ATTR_CURRENT_VERSION,
+    ATTR_IN_PROGRESS,
+    ATTR_LATEST_VERSION,
+    ATTR_RELEASE_SUMMARY,
+    ATTR_RELEASE_URL,
+    ATTR_SKIPPED_VERSION,
+    ATTR_TITLE,
+    ATTR_VERSION,
+    DOMAIN,
+    SERVICE_INSTALL,
+    SERVICE_SKIP,
+    SUPPORT_BACKUP,
+    SUPPORT_INSTALL,
+    SUPPORT_PROGRESS,
+    SUPPORT_SPECIFIC_VERSION,
+)
+
+SCAN_INTERVAL = timedelta(minutes=15)
+MIN_TIME_BETWEEN_SCANS = timedelta(seconds=10)
+
+ENTITY_ID_FORMAT: Final = DOMAIN + ".{}"
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class UpdateDeviceClass(StrEnum):
+    """Device class for update."""
+
+    FIRMWARE = "firmware"
+
+
+DEVICE_CLASSES_SCHEMA = vol.All(vol.Lower, vol.Coerce(UpdateDeviceClass))
+
+
+__all__ = [
+    "ATTR_BACKUP",
+    "ATTR_VERSION",
+    "DEVICE_CLASSES_SCHEMA",
+    "DOMAIN",
+    "PLATFORM_SCHEMA_BASE",
+    "PLATFORM_SCHEMA",
+    "SERVICE_INSTALL",
+    "SERVICE_SKIP",
+    "SUPPORT_BACKUP",
+    "SUPPORT_INSTALL",
+    "SUPPORT_PROGRESS",
+    "SUPPORT_SPECIFIC_VERSION",
+    "UpdateDeviceClass",
+    "UpdateEntity",
+    "UpdateEntityDescription",
+]
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up Select entities."""
+    component = hass.data[DOMAIN] = EntityComponent(
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+    await component.async_setup(config)
+
+    component.async_register_entity_service(
+        SERVICE_INSTALL,
+        {
+            vol.Optional(ATTR_VERSION): cv.string,
+            vol.Optional(ATTR_BACKUP): cv.boolean,
+        },
+        async_install,
+    )
```

Comment:
You can pass required features to the register entity service.

```suggestion
        SUPPORT_INSTALL,
    )
```

Reply:
Hmm was looking at our code in helpers/service.py that deals with this actually won't throw an error. I still think that we should register the supported feature flag here, and update the helper to raise if entity was explicitly referenced.

https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/service.py#L641

Reply:
PR with raise https://github.com/home-assistant/core/pull/68394

Reply:
I guess we really going to need an option to allow a service call to fail :)

Reply:
Adjusted in [`777e5ba` (#68248)](https://github.com/home-assistant/core/pull/68248/commits/777e5ba88f716106821a2b95478cbc0adc9c3352)